File: 0666-path-sum-iv/0666-path-sum-iv.py
class Solution:
    def pathSum(self, nums: List[int]) -> int:
        sum, tree = 0, {}
        for x in nums: tree[x//10] = x%10
    
        def dfs(root, pre):
            nonlocal sum
            cur = tree[root] + pre
            lev, pos = root//10, root%10
            left, right = (lev+1)*10+2*pos-1, (lev+1)*10+2*pos
            if left not in tree and right not in tree:
                sum += cur
                return
            if left in tree: dfs(left, cur)
            if right in tree: dfs(right, cur)

        dfs(nums[0]//10, 0)
        return sum

File: 0666-path-sum-iv/0666-path-sum-iv.java
class Solution {
    int sum = 0;
    Map<Integer,Integer> tree = new HashMap<>();
    public int pathSum(int[] nums) {
        if (nums == null || nums.length == 0) return 0;

        for (int num : nums) {
            int key = num / 10;
            int value = num % 10;
            tree.put(key, value);
        }
        dfs(nums[0]/10, 0);
        return sum;
    }
    
    private void dfs(int root, int pre) {
        int cur = pre + tree.get(root);
        int lev = root/10, pos = root%10;
        int left = ++lev*10+pos*2-1, right = lev*10+pos*2;
        if(!tree.containsKey(left) && !tree.containsKey(right)) {
            sum += cur;
            return;
        }
        if(tree.containsKey(left)) dfs(left, cur);
        if(tree.containsKey(right)) dfs(right, cur);
    }
}